### PR TITLE
Cleanup lpstream, Connection and BufferStream

### PR DESCRIPTION
**Pull request by dryajov**
This PR further cleans up and simplifies stream management. It removes the nested stream from `Connection` and instead relies on dynamic dispatch to properly propagate stream closing. 

In addition to this the PR also adds better object tracing and metrics. This is still a wip and would need to be rethought at a later point, but it allows me to further troubleshoot proper object destruction and cleanup.

### CODE REVIEW

**Comment by dryajov** [comment]
can I get some eyes on this please @arnetheduck @sinkingsugar

File: libp2p/muxers/mplex/mplex.nim [hunk 1]
```diff
@@ -105,22 +107,16 @@ method handle*(m: Mplex) {.async, gcsafe.} =
...
                 try:
-                  await m.streamHandler(stream)
+                  await m.streamHandler(channel)
                   trace "finished handling stream"
-                  # doAssert(stream.closed, "connection not closed by handler!")
+                  # doAssert(channel.closed, "connection not closed by handler!")
                 except CatchableError as exc:
                   trace "exception in stream handler", exc = exc.msg
-                  doAssert(stream.closed, "stream not closed by protocol handler")
+                  doAssert(channel.closed, "stream not closed by protocol handler")
```
Comment (sinkingsugar) [inline libp2p/muxers/mplex/mplex.nim:118]: this finally works fine now btw?

File: libp2p/protocols/secure/secure.nim [hunk 2]
```diff
@@ -11,15 +11,27 @@ import options
...
     buf: StreamSeq
 
+method initStream*(s: SecureConn) =
+  if s.objName.len == 0:
+    s.objName = "SecureConn"
+
+  procCall Connection(s).initStream()
+
+method close*(s: SecureConn) {.async.} =
+  await s.stream.close()
```
Comment (sinkingsugar) [inline libp2p/protocols/secure/secure.nim:31]: is stream surely not nil?

File: libp2p/stream/lpstream.nim [hunk 3]
```diff
@@ -7,20 +7,21 @@
...
 
 type
   LPStream* = ref object of RootObj
     isClosed*: bool
     isEof*: bool
-    closeEvent*: AsyncEvent
-    when chronicles.enabledLogLevel == LogLevel.TRACE:
-      oid*: Oid
+    objName*: string
+    oid*: Oid
```
Comment (sinkingsugar) [inline libp2p/stream/lpstream.nim:24]: still not sure why we need `oid` vs `cast[int](self)` since its a `ref`

**Comment by sinkingsugar** [comment]
If I had to choose how to refactor I would have personally refactored using composition rather then inheritance ( I see Connection super class proliferated ), but it's not a big deal at all. 
Looks like a good refactor to me.

File: libp2p/protocols/secure/secure.nim (hunk 2 above)
Comment (dryajov) [reply libp2p/protocols/secure/secure.nim:31]: good point, you never know with async, so good idea to guard against it.

**Comment by dryajov** [comment]
> If I had to choose how to refactor I would have personally refactored using composition rather then inheritance ( I see Connection super class proliferated ), but it's not a big deal at all.
> Looks like a good refactor to me.

honestly that's what I'm moving away from in this PR, the reason being because there isn't any sane way to propagate stream closing/closure upstream. 

Consider this example:

- Class `A -> B` and `A -> C`, where C is a facade of sorts that wraps around `B`
- Now, when I want to propagate messages from `C` to the member it wraps, I can do `C.field.foo()`
- But, when the wrapped `B` wants to propagate something to `C`, there is no way, ie there is no `B.parent.foo()`
  - because that would create a circular reference and possibly lead to memory leaks in subtle hard to troubleshoot ways
  - besides it's a leaking abstraction because we now have to expose our internal object structure/hierarchy for no good reason
- Inheritance solves this because it guarantees that the correct method will be dispatched (ie polymorphism). Moreover, you now only have one object to manage instead of two - the wrapper and wrappee.

This is the issue with the previous approach where i used composition instead. Specifically in mplex, where closing a channel would not close the parent `Connection` because the wrapped object doesn't know anything about it's parent (and it shouldn't either). 

To illustrate this here is a brief summary of the flow:

- A `Connection` wrapps an `LPStream` stream, which in the muxer is an `LPChannel`
- User protocol handlers would only interact with the `Connection`
- The muxer however, needs to have access to `LPChannel`, because it exposes additional functionality, such as channel `id`, `pushTo`, `closeRemote` and `reset` methods which don't exist in `Connection`
- The muxer then needs to keep both, the `LPChannel` and `Connection` objects around
- When a request arrives the muxer interacts directly with `LPChannel` and either pushes data onto it or close/resets it
- When a protocol `handler` calls `Connection.close`, I can dispatch that easily, because `close` would internally dispatch it to the wrapped `LPChannel`
- However, when the channel is being closed externally, ie, when the remote peer calls `close` or `reset`, I now need to dispatch it to both objects
  - In the case of `close`, it isn't as problematic, as I can just call `Connection.close` and that would take care of it
  - But in the case of `reset`, this breaks, because a) there is no `Connection.reset`, so I can't just call the wrapper; b) if I call `LPChannel.reset` I then still need to call `Connection.close` because otherwise the two object will be in an inconsistent state, and this is the problem, because calling `Connection.close` would also trigger `LPStream.close`, which isn't at all what we need

The disadvantage of this approach is that I now have to keep two objects around and make sure that neither goes out of sync; carefully track that all calls propagate up and down properly, etc... In short, it complicates things for no good reason.

I hope this clarifies it.

File: libp2p/muxers/mplex/mplex.nim (hunk 1 above)
Comment (dryajov) [reply libp2p/muxers/mplex/mplex.nim:118]: actually no, we should probably reset the channel there either way

File: libp2p/stream/lpstream.nim (hunk 3 above)
Comment (dryajov) [reply libp2p/stream/lpstream.nim:24]: I was actually thinking about it, probably better to just use `hash()` either way, whatever it ends up being under the hood.

Comment (dryajov) [reply libp2p/stream/lpstream.nim:24]: and probably a good idea to have a `hash()` implemented for most objects either way

Comment (sinkingsugar) [reply libp2p/stream/lpstream.nim:24]: Yes hash will use the pointer just like the cast iirc!

**Comment by sinkingsugar** [comment]
Well you are right, the lack of proper first class weak references in nim is also the reason arc will likely always leak too!

The thing is that also methods feel fragile to me language wise...

But yes this is probably the best solution that does not involve writing a DSL to fix the lack of proper tools :)

File: libp2p/stream/lpstream.nim (hunk 3 above)
Comment (dryajov) [reply libp2p/stream/lpstream.nim:24]: I'll rework it in the next pr, don't want this one to go stale...

File: libp2p/muxers/muxer.nim [hunk 4]
```diff
@@ -65,3 +66,6 @@ method init(c: MuxerProvider) =
       trace "exception in muxer handler", exc = exc.msg
 
   c.handler = handler
+
+proc `$`*(m: Muxer): string =
+  $m.connection.peerInfo
```
Comment (arnetheduck) [inline libp2p/muxers/muxer.nim:71]: peerinfo may be nil

File: libp2p/stream/connection.nim [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+proc setupConnectionTracker(): ConnectionTracker =
+  result = new ConnectionTracker
+  result.opened = 0
+  result.closed = 0
+  result.dump = dumpTracking
+  result.isLeaked = leakTransport
+  addTracker(ConnectionTrackerName, result)
+
+proc init*[T: Connection](self: var T, peerInfo: PeerInfo): T =
```
Comment (arnetheduck) [inline libp2p/stream/connection.nim:54]: where does this peerinfo go?

**Comment by arnetheduck** [comment]
we've brought up the propagation in different PR's - in general, when propagations collide, issues ensue - ie when you have propagations in both directions, things become difficult: this is the difficulty you feel when doing composition.

A way to look at the issue is that composition forces you to be explicit about propagation which in turn makes it "onerous" to do dubious propagation patterns and "forces" you to come up with a strategy that has a more simple flow.

Of course, you can hide behind methods, dynamic dispatch etc, hoping that at runtime things resolve themselves as you intended in the design, though this is inherently a weaker model that relies more on hopefulness that the developer gets it right.

File: libp2p/stream/connection.nim (hunk 5 above)
Comment (dryajov) [reply libp2p/stream/connection.nim:54]: good catch

**Comment by dryajov** [comment]
> A way to look at the issue is that composition forces you to be explicit about propagation which in turn makes it "onerous" to do

I'd really love to see an example of how that would be done. No sarcasm, I'd serious love to see something more tangible.

One reason that makes me doubt about the statement is the fact that `pipelines` (as a pattern in streams) exist explicitly to solve the issue of up and down state propagation, which in turn enable composition. The more imperative style of streams we've chosen here (which is not without its merits) unfortunately lacks that ability almost inherently.

> dubious propagation  patterns and "forces" you to come up with a strategy that has a more simple flow.

I'm all for simplifying the flow, let's work on making it happen.

**Comment by dryajov** [comment]
Btw, I don't mean it like a full blown PR or anything, some pseudo code would suffice. 

We really should have this discussion and I don't see a better way of kicking it of, but a proposal of sorts.

**Pull request merged**